### PR TITLE
Initial refactoring work

### PR DESCRIPTION
**Pull request by olo-ntaylor**

### CODE REVIEW

File: src/Thoth.Json.Net.fsproj [hunk 1]
```diff
@@ -8,6 +8,7 @@
     <PackageReadmeFile>README.md</PackageReadmeFile>
     <PackageTags>fable;fable-library;fable-dotnet;fsharp;json</PackageTags>
     <Authors>Maxime Mangel</Authors>
+    <TargetFramework>net6.0</TargetFramework>
```
Comment (jfaixolo) [inline src/Thoth.Json.Net.fsproj:11]: ❓ Was there any particular reason to switch from Standard to .NET 6? As a package, Standard would give consumers more flexibility. (I think we can safely drop framework 4.6 though.)

File: tests/Tests.fsproj [hunk 2]
```diff
@@ -1,35 +1,27 @@
...
+    <Compile Include="ExtraCoders.fs" />
     <Compile Include="Main.fs" />
   </ItemGroup>
-  <Import Project="..\.paket\Paket.Restore.targets" />
+  <ItemGroup>
+    <PackageReference Update="FSharp.Core" Version="6.0.7" />
+    <PackageReference Update="Expecto">
+      <Version>9.0.4</Version>
+    </PackageReference>
+    <PackageReference Update="Mono.Cecil">
```
Comment (jfaixolo) [inline tests/Tests.fsproj:23]: ❓ Is Cecil needed if we drop the reflection-based features?

File: src/Types.fs [hunk 3]
```diff
@@ -21,32 +21,31 @@ type CaseStrategy =
 
 type DecoderError = string * ErrorReason
 
-type Decoder<'T> = string -> JsonValue -> Result<'T, DecoderError>
+type Decoder<'T> = string -> JsonElement -> Result<'T, DecoderError>
 
-type Encoder<'T> = 'T -> JsonValue
+type Encoder<'T> = 'T -> JsonNode
```
Comment (jfaixolo) [inline src/Types.fs:26]: ❓ What is the distinction between `JsonElement`/`JsonNode`?

File: src/Thoth.Json.Net.fsproj (hunk 1 above)
Comment (olo-ntaylor) [reply src/Thoth.Json.Net.fsproj:11]: That's a good call-out; I wasn't really thinking but yeah targeting standard makes more sense 👍

File: tests/Tests.fsproj (hunk 2 above)
Comment (olo-ntaylor) [reply tests/Tests.fsproj:23]: I don't even know what it is haha, but I'll check. Probably not, though, based on some quick googling it looks like it has to do with inspecting IL code/reflection

File: src/Types.fs (hunk 3 above)
Comment (olo-ntaylor) [reply src/Types.fs:26]: So, the `System.Text.Json` library exposes two different DOMs. One is optimized for reading-- that's `JsonDocument`/`JsonElement`. And one is optimized for writing/creation/mutation -- that's `JsonNode`/`JsonValue`.

So, for _decoding_, we want to use the read-optimized/immutable `JsonElement`. And for _encoding_, we want to (have to) use the write-optimized/mutable `JsonNode`/`JsonValue`.

**Pull request closed without merging**